geomet/esri.py

```diff
@@ -0,0 +1,230 @@
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import json
+import geomet
+
+def load(source_file):
+    """
+    Converts Esri Json File to GeoJSON.
+
+    :param source_file:
+        Path to a file that contains the Esri JSON data.
+    
+    :returns: 
+         A GeoJSON `dict` representing the geometry read from the file.
+
+    """
+    with open(source_file, 'r') as reader:
+        data = json.loads(reader.read())
+        return loads(data)
+
+def loads(string):
+    """
+    Construct a GeoJSON `dict` from Esri JSON (string/dict).
+
+    :param string:
+        The Esri JSON geometry representation
+
+    :returns: 
+         A GeoJSON `dict` representing the geometry read from the file.
+    """
+    if isinstance(string, str):
+        string = json.loads(string)
+    if 'rings' in string:
+        return _esri_to_geojson_convert['rings'](string)
+    elif 'paths' in string:
+        return _esri_to_geojson_convert['paths'](string)
+    elif 'x' in string or 'y' in string:
+        return _esri_to_geojson_convert['x'](string)
+    elif 'points' in string:
+        return _esri_to_geojson_convert['points'](string)
+    else:
+        raise geomet.InvalidGeoJSONException('Invalid EsriJSON: %s' % string)
+
+        
+
+def dump(obj, dest_file):
+    """
+    Converts GeoJSON to Esri JSON File.
+    """
+    with open(dest_file, 'w') as writer:
+        writer.write(json.dumps(dumps(obj)))
+    return dest_file
+
+def dumps(obj):
+    """
+    Dump a GeoJSON-like `dict` to a Esri JSON.
+    """
+    if 'type' in obj and \
+        obj['type'].lower() in _gj_to_esri.keys():
+        convert = _gj_to_esri[obj['type'].lower()]
+        return convert(obj)
+    else:
+        raise geomet.InvalidGeoJSONException("Invalid GeoJSON type %s" % obj)
+
+def _extract_geojson_srid(obj):
+    """
+    Extracts the SRID code (WKID code) from geojson. If not found, SRID=4326
+
+    :returns: Integer
+    """
+    meta_srid = obj.get('meta', {}).get('srid', None)
+    # Also try to get it from `crs.properties.name`:
+    crs_srid = obj.get('crs', {}).get('properties', {}).get('name', None)
+    if crs_srid is not None:
+        # Shave off the EPSG prefix to give us the SRID:
+        crs_srid = crs_srid.replace('EPSG', '')
+
+    if (meta_srid is not None and
+            crs_srid is not None and
+            str(meta_srid) != str(crs_srid)):
+        raise ValueError(
+            'Ambiguous CRS/SRID values: %s and %s' % (meta_srid, crs_srid)
+        )
+    srid = meta_srid or crs_srid
+
+    return srid or 4326
```

Comment:
If the EsriJSON spec requires an SRID to be specified, then it should be an error if it's not. Defaulting to WGS84 will often be wrong.

Reply:
According to the GeoJSON spec, the default SRID is 4326 or WGS-1984: ```Default coordinate reference system: The default is a geographic coordinate reference system, using the WGS84 datum, and with longitude and latitude units of decimal degrees.``` so this is the correct way to handle this.

Reply:
This feels different to me because we're adding data to the geometry without knowing if it's right or not, and it's data which will cause the geometry to be read improperly if it's wrong, so I would personally prefer err on the side of caution and raise an error over generating data from nowhere. I'd love to hear from others on this though. Perhaps @larsbutler or @sgillies could chime in with some thoughts.

Reply:
The value can be null on Esri JSON, so this isn't an issue but since crs or spatial references are optional, it will probably produce many EsriJSON without a coordinate reference system.  @sgillies thoughts?

Reply:
I agree with @tomplex on this one; if there isn't a way to preserve important context when converting (this is huge problem, I find, with many of these data formats), I don't think it's a good idea to make assumptions about the SRID.

@achapkowski  As an alternative, what about adding an optional parameter to pass into the `load/loads` function for a `default_srid` or something like that? This way, the user can control which value gets assigned and it reduces the possibility for mistakes.

Reply:
@larsbutler sounds good, I'll add the new paramter.

Reply:
I'm 👍 on the new parameter, but I'm still not comfortable defaulting to 4326 if it's not specified by the user one way or another. I'd rather it be `None` or have an error raised. Even if the input is WGS84 99% of the time, I'd rather force them to provide it then assume it and be the cause of a bug further downstream in their data processing.

Reply:
Ok, I'll set it to ```null``` if not provided.

Reply:
Thanks @achapkowski  - sorry to be such a pain here but I want to make sure we get it right.